server/src/main/java/org/opensearch/wlm/QueryGroupService.java

```diff
@@ -8,32 +8,194 @@
+    }
+
+    /**
+     * run at regular interval
+     */
+    protected void doRun() {
+        if (workloadManagementSettings.getWlmMode() == WlmMode.DISABLED) {
+            return;
+        }
+        taskCancellationService.refreshQueryGroups(activeQueryGroups, deletedQueryGroups);
+        taskCancellationService.cancelTasks(nodeDuressTrackers::isNodeInDuress);
+        taskCancellationService.pruneDeletedQueryGroups();
+    }
+
+    private QueryGroupState getQueryGroupState(final String queryGroupId) {
+        return queryGroupStateMap.get(queryGroupId);
+    }
+
+    /**
+     * {@link AbstractLifecycleComponent} lifecycle method
+     */
+    @Override
+    protected void doStart() {
+        scheduledFuture = threadPool.scheduleWithFixedDelay(() -> {
+            try {
+                doRun();
+            } catch (Exception e) {
+                logger.debug("Exception occurred in Query Sandbox service", e);
+            }
+        }, this.workloadManagementSettings.getQueryGroupServiceRunInterval(), ThreadPool.Names.GENERIC);
+    }
+
+    @Override
+    protected void doStop() {
+        if (scheduledFuture != null) {
+            scheduledFuture.cancel();
+        }
+    }
+
+    @Override
+    protected void doClose() throws IOException {}
+
+    @Override
+    public void applyClusterState(ClusterChangedEvent event) {
+        // Retrieve the current and previous cluster states
+        Metadata previousMetadata = event.previousState().metadata();
+        Metadata currentMetadata = event.state().metadata();
+
+        // Extract the query groups from both the current and previous cluster states
+        Map<String, QueryGroup> previousQueryGroups = previousMetadata.queryGroups();
+        Map<String, QueryGroup> currentQueryGroups = currentMetadata.queryGroups();
+
+        // Detect new query groups added in the current cluster state
+        for (String queryGroupName : currentQueryGroups.keySet()) {
+            if (!previousQueryGroups.containsKey(queryGroupName)) {
+                // New query group detected
+                QueryGroup newQueryGroup = currentQueryGroups.get(queryGroupName);
+                // Perform any necessary actions with the new query group
+                this.activeQueryGroups.add(newQueryGroup);
+                queryGroupStateMap.put(newQueryGroup.get_id(), new QueryGroupState());
+            }
+        }
+
+        // Detect query groups deleted in the current cluster state
+        for (String queryGroupName : previousQueryGroups.keySet()) {
+            if (!currentQueryGroups.containsKey(queryGroupName)) {
+                // Query group deleted
+                QueryGroup deletedQueryGroup = previousQueryGroups.get(queryGroupName);
+                // Perform any necessary actions with the deleted query group
+                this.deletedQueryGroups.add(deletedQueryGroup);
+                queryGroupStateMap.remove(deletedQueryGroup.get_id());
+            }
+        }
```

Comment:
I will disagree to this. Since according to the method definition
```
public Set<String> changedCustomMetadataSet() {
        Set<String> result = new HashSet<>();
        Map<String, Metadata.Custom> currentCustoms = state.metadata().customs();
        Map<String, Metadata.Custom> previousCustoms = previousState.metadata().customs();
        if (currentCustoms.equals(previousCustoms) == false) {.    ===========> this already does iterate on all custom metadata types in worst case
            for (Map.Entry<String, Metadata.Custom> currentCustomMetadata : currentCustoms.entrySet()) {
                // new custom md added or existing custom md changed
                if (previousCustoms.containsKey(currentCustomMetadata.getKey()) == false
                    || currentCustomMetadata.getValue().equals(previousCustoms.get(currentCustomMetadata.getKey())) == false) {
                    result.add(currentCustomMetadata.getKey());
                }
            }
            // existing custom md deleted
            for (Map.Entry<String, Metadata.Custom> previousCustomMetadata : previousCustoms.entrySet()) {
                if (currentCustoms.containsKey(previousCustomMetadata.getKey()) == false) {
                    result.add(previousCustomMetadata.getKey());
                }
            }
        }
```
even though we add something like

```
if (!event.changedCustomMetadataSet().contains(QueryGroupMetadata.TYPE)) return;
```
I highly doubt this will result in better performance. On top of it the mentioned method is not consumed anywhere in the code by other `ClusterStateApplier`s.